### PR TITLE
[5.5] Add stream download response

### PR DESCRIPTION
**Pull request by TheoKouzelis**
Adds the ability to respond with remote files as downloads. 

```
<?php

    Route::get('/', function () {
        return response()->streamDownload(function () {
            echo file_get_contents('https://my.remote.com/file/store-12345.jpg');
        }, 'nice-name.jpg');
    });
```

### CODE REVIEW

File: src/Illuminate/Routing/ResponseFactory.php [hunk 1]
```diff
@@ -111,6 +111,33 @@ public function stream($callback, $status = 200, array $headers = [])
         return new StreamedResponse($callback, $status, $headers);
     }
 
+    /**
+     * Return a new streamed response as a file download from the application.
+     *
+     * @param  \Closure  $callback
+     * @param  string  $name
```
Comment (mfn) [inline src/Illuminate/Routing/ResponseFactory.php]: `string|null` according to your method signature

**Review by mfn — CHANGES_REQUESTED** [review]
No tests?

File: src/Illuminate/Routing/ResponseFactory.php (hunk 1 above)
Comment (TheoKouzelis) [reply src/Illuminate/Routing/ResponseFactory.php]: Hi @mfn 

Thanks for the PR review. I have pretty much copied the doc block from the "download" method. Would you also like me to make the same change to that doc block as well?

**Comment by TheoKouzelis** [comment]
@mfn I couldn't find any tests for this class to add to. Do you know where a test for this method should go?

**Comment by sisve** [comment]
1. This looks very similar to the existing stream() method. Can't it be used for this purpose?
2. Should this new method be part of the ResponseFactory contract?

**Comment by TheoKouzelis** [comment]
Hi @sisve 

Thanks for the PR review.

I was trying to copy the feel of the existing API of "[file](https://github.com/TheoKouzelis/framework/blob/3e5fcc17055f10783c053e74cb55d860d4285ed7/src/Illuminate/Routing/ResponseFactory.php#L179)" and "[download](https://github.com/TheoKouzelis/framework/blob/3e5fcc17055f10783c053e74cb55d860d4285ed7/src/Illuminate/Routing/ResponseFactory.php#L150)". Where file is just a basic wrapper around the Symfony class BinaryFileResponse and download goes one step further and sets up the content disposition. This PR was kind of born out of this [stackoverflow question](https://stackoverflow.com/questions/38791635/download-a-file-in-laravel-using-a-url-to-external-resource/) where it felt like people where searching for a similar function to download for remote files.

Good point about the ResponseFactory contract. I have missed that and I will get this added to this PR tonight.

**Comment by TheoKouzelis** [comment]
@m1guelpf OK cool thanks. Should I remove this PR and make a new branch of 5.6?

**Comment by mfn** [comment]
Yes, thanks!

**Comment by TheoKouzelis** [comment]
What is the discussion process behind adding a method to a contract? Because I'm looking at the [ResponseFactory](https://github.com/laravel/framework/blob/5.5/src/Illuminate/Contracts/Routing/ResponseFactory.php) in more detail and I can see "stream" and "download" are part of the interface but not "file". So it now makes me wonder if this method should be?

**Comment by TheoKouzelis** [comment]
This has now been moved to a [PR on the 5.6 branch](https://github.com/laravel/framework/pull/22777) with @mfn doc block change and @sisve contract change added

**Pull request closed without merging**